cdisc_rules_engine/operations/base_operation.py

```diff
@@ -0,0 +1,51 @@
+import pandas as pd
+from cdisc_rules_engine.models.operation_params import OperationParams
+from abc import abstractmethod
+
+from cdisc_rules_engine.services.cache.cache_service_interface import (
+    CacheServiceInterface,
+)
+from cdisc_rules_engine.services.data_services.base_data_service import BaseDataService
+
+
+class BaseOperation:
+    def __init__(
+        self,
+        params: OperationParams,
+        original_dataset: pd.DataFrame,
+        cache_service: CacheServiceInterface,
+        data_service: BaseDataService,
+    ):
+        self.params = params
+        self.cache = cache_service
+        self.data_service = data_service
+        self.evaluation_dataset = original_dataset
+
+    @abstractmethod
+    def _execute_operation(self):
+        """Perform operation calculations."""
+        pass
+
+    def execute(self) -> pd.DataFrame:
+        result = self._execute_operation()
+        return self._handle_operation_result(result)
+
+    def _handle_operation_result(self, result) -> pd.DataFrame:
+        if self.params.grouping:
+            # Handle grouped results
+            result = result.rename(
```

Comment:
Maybe we can move grouping to separate method to clear handling method

Reply:
updated